pkg/remote/builtin/gitlab/gitlab.go

```diff
@@ -259,20 +230,66 @@ func (r *Gitlab) Hook(req *http.Request) (*common.Hook, error) {
 		return nil, err
 	}
 
-	if len(parsed.After) == 0 || parsed.TotalCommitsCount == 0 {
-		return nil, nil
-	}
+	switch parsed.ObjectKind {
+	case "merge_request":
+		if parsed.ObjectAttributes.State != "reopened" && parsed.ObjectAttributes.MergeStatus != "unchecked" ||
+			parsed.ObjectAttributes.State != "opened" && parsed.ObjectAttributes.MergeStatus != "unchecked" {
+			return nil, nil
+		}
+
+		return mergeRequest(parsed, req)
+	case "tag_push", "push":
+		if len(parsed.After) == 0 || parsed.TotalCommitsCount == 0 {
+			return nil, nil
+		}
 
-	if parsed.ObjectKind == "merge_request" {
-		// NOTE: in gitlab 8.0, gitlab will get same MR models as github
-		//       https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/981/diffs
+		return push(parsed, req)
+	default:
 		return nil, nil
 	}
+}
 
-	if len(parsed.After) == 0 {
-		return nil, nil
+func mergeRequest(parsed *gogitlab.HookPayload, req *http.Request) (*common.Hook, error) {
+	var hook = new(common.Hook)
+
+	re := regexp.MustCompile(".git$")
+
+	hook.Repo = &common.Repo{}
+	hook.Repo.Owner = req.FormValue("owner")
+	hook.Repo.Name = req.FormValue("name")
+	hook.Repo.FullName = fmt.Sprintf("%s/%s", hook.Repo.Owner, hook.Repo.Name)
+	hook.Repo.Link = re.ReplaceAllString(parsed.ObjectAttributes.Target.HttpUrl, "$1")
+	hook.Repo.Clone = parsed.ObjectAttributes.Target.HttpUrl
+	hook.Repo.Branch = "master"
+
+	hook.Commit = &common.Commit{}
+	hook.Commit.Message = parsed.ObjectAttributes.LastCommit.Message
+	hook.Commit.Sha = parsed.ObjectAttributes.LastCommit.Id
+	hook.Commit.Remote = parsed.ObjectAttributes.Source.HttpUrl
+
+	if parsed.ObjectAttributes.Source.HttpUrl == parsed.ObjectAttributes.Target.HttpUrl {
+		hook.Commit.Ref = fmt.Sprintf("refs/heads/%s", parsed.ObjectAttributes.SourceBranch)
+		hook.Commit.Branch = parsed.ObjectAttributes.SourceBranch
+		hook.Commit.Timestamp = time.Now().UTC().Format("2006-01-02 15:04:05.000000000 +0000 MST")
+	} else {
+		hook.Commit.Ref = fmt.Sprintf("refs/merge-requests/%d/head", parsed.ObjectAttributes.IId)
+		hook.Commit.Branch = parsed.ObjectAttributes.SourceBranch
+		hook.Commit.Timestamp = time.Now().UTC().Format("2006-01-02 15:04:05.000000000 +0000 MST")
```

Comment:
Also, I don't think this and the line above need to be in both branches of the if/else statement.